Add sound mode support (media_player frontend)

starkillerOG opened the pull request:
Added frontend support for an drop down box select for the sound mode (also updated with status).
This is general support for all media_player components.
This box is not shown in media_players that did not (yet) implement sound mode support in the backend.
I implemented the sound mode support (backend) for the denonavr platform and that one works perfictly.
I will make a pull request for the backend soon (still dooing some cleanup).

I tested this code and it works on Hassbian on my pi.

On the home assistant chat I heard that other platforms like onkyo and sonos could also implement sound mode support and probably most receivers could support this.

Backend pull request:
https://github.com/home-assistant/home-assistant/pull/13706
Documentation pull request:
https://github.com/home-assistant/home-assistant.github.io/pull/4470

starkillerOG committed:
Add sound mode support to media_player (frontend)

- this is part of the added sound mode support to the home assistant backend

starkillerOG committed:
Add sound mode support

Changed icon of the new input select.
Tested the code and it works, for other media players that did not implement sound mode support the input select is nicely hidden (not present).
For the Denonavr platform it works and sound mode is updated and can be changed in the front pannel.

starkillerOG commented:
![sound_mode_support](https://user-images.githubusercontent.com/14811189/35101794-1669b1b4-fc61-11e7-8d61-337fe2c49fbe.png)

starkillerOG commented:
![sound_mode_support_2](https://user-images.githubusercontent.com/14811189/35101859-550ba044-fc61-11e7-96a8-0401984fc9ef.png)

starkillerOG commented:
@balloob @c727 Could one of you maybe revieuw this PR and then merge it?

c727 commented on src/more-infos/more-info-media_player.html:
```diff
@@ -314,6 +357,14 @@
+      return isOff || !supportsSelectSoundMode;
+    }
+
+    computeSelectedSoundMode(stateObj) {
```
Not used

starkillerOG replied:
I have removed it

c727 commented on src/more-infos/more-info-media_player.html:
```diff
@@ -45,6 +45,15 @@
         margin-left: 10px;
       }
 
+      iron-icon.sound-mode-input {
```
Same as source-input. You should combine them

starkillerOG replied:
I combined both

starkillerOG committed:
Add sound mode support

- combine style classes of sound mode and source.
- remove unused computeSelectedSoundMode

starkillerOG commented:
@c727 I commited the changes you proposed, can you now merge this PR?

starkillerOG commented:
@c727 I tested the new changes and it seems to work with those unused parts of the code removed.

c727 commented:
I can't do this, you have to wait for Andrey, armills or balloob

starkillerOG commented:
@andrey-git @armills @balloob, could one of you merge this PR?

andrey-git commented on src/more-infos/more-info-media_player.html:
```diff
@@ -106,6 +106,18 @@
       </div>
+      <!-- SOUND MODE PICKER -->
+      <div class='controls layout horizontal justified'
+           hidden$='[[computeHideSelectSoundMode(isOff, supportsSelectSoundMode)]]'>
```
dom-if it instead

starkillerOG replied:
I don't know exactly what you mean but I guess you mean to use a "dom-if" property in the <div> instead of using the hidden$.
Again I am not an expert in how this works and what the diffrence is.
But the rest of the file is using the hidden$ property.

Basically I made a copy of the Source input, so everything is done in a similar way as the already implemented Source input.

Would it not be better to keep everything the same as the rest of the file for now.
Maybe someone else who does no what the diffrence is and nows how to use these properties correctly can change it for the whole file at once (also for the SourceInput, Power and Volume).

andrey-git commented on src/more-infos/more-info-media_player.html:
```diff
@@ -359,6 +397,27 @@
     }
 
+    handleSoundModeChanged(soundModeIndex, soundModeIndexOld) {
+      var soundModeInput;
```
No need to predeclare variables.

Just use `const` when you assign it.

starkillerOG replied:
All handle's are done with var instead of const throught this file.
honestly I don't know the exact diffrence but I think it might be better to keep the file uniform and do the same thing as done in the rest of the file.

Basically I made a copy of the Source input, so everything is done in a similar way as the already implemented Source input.

starkillerOG commented:
@andrey-git Do you agree to keep the document uniform?
Or shall I try to implement the sugested changes (const and dom-if)?

starkillerOG commented:
@andrey-git Shall I try to implement the suggested changes or do you agree with my previous comment?

starkillerOG committed:
Add sound mode support

Implemented changes suggested by @andrey-git

balloob commented:
Closed as new supported features need to be discussed in the architecture repo first as per https://developers.home-assistant.io/docs/en/entity_index.html#changing-the-entity-model
